product_supplierinfo_for_customer/models/product_product.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2015 OdooMRP team
+# Copyright 2015 AvanzOSC
+# Copyright 2015 Tecnativa
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import api, models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    @api.multi
+    def _get_price_from_supplierinfo(self, partner_id):
+        if partner_id:
+            for product in self:
+                supplierinfo = self.env['product.supplierinfo'].search(
+                    ['|', ('product_tmpl_id', '=', product.product_tmpl_id.id),
+                     ('product_id', '=', product.id),
+                     ('type', '=', 'customer'),
+                     ('name', '=', partner_id)])
+                if supplierinfo:
+                    return supplierinfo.price
+        return 0.0
+
+    @api.multi
+    def price_compute(
+            self, price_type, uom=False, currency=False, company=False):
+        if price_type == 'partner':
+            partner = self.env.context.get('partner_id', False) or \
+                self.env.context.get('partner', False)
+            for product in self:
+                price = product._get_price_from_supplierinfo(partner)
+                if not price:
+                    return super(ProductProduct, self).price_compute(
+                        'list_price', uom, currency, company)
+                prices = dict.fromkeys(self.ids, 0.0)
```

Comment:
:stop_sign:  you are overwriting the dict on every iteration
as well as you will only have one iteration, as you return `prices` inside the cycle